### PR TITLE
Add a CostFactory and Get auto routes to work

### PR DESCRIPTION
**Pull request by dnesbitt61**
Fixed a couple issues with PathAlgorithm (was calling wrong A* method!) and fix some pure virtual methods in Bicycle and Auto costing.

@dgearhart @gknisely @kevinkreiser

### CODE REVIEW

**Comment by gknisely** [comment]
ship it

File: valhalla/thor/costfactory.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+#include "thor/bicyclecost.h"
+#include "thor/pedestriancost.h"
+
+namespace valhalla {
+namespace thor {
+
+/**
+* Generic factory class for creating objects based on type name.
+*/
+template <typename T>
```
Comment (kevinkreiser) [inline valhalla/thor/costfactory.h]: whats the template for?

File: valhalla/thor/costfactory.h [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+  /**
+   * Register
+   */
+  CostFactory& Register(const std::string& type, DynamicCost* obj) {
+    factory_funcs_.insert(std::make_pair(type, obj));
+    return (*this);
+  }
+
+  DynamicCost* Create(const std::string& type) const {
```
Comment (kevinkreiser) [inline valhalla/thor/costfactory.h]: why must these be pointers? also if they must be pointers can we at least use shared_ptr so we dont have to worry about deallocating/copying them?

Comment (dnesbitt61) [inline valhalla/thor/costfactory.h]: Yea, maybe you can help me with shared_ptr. I tried that (and the template was probably leftover from another Factory model I tried). I'm going to merge this and we can rework it later.

**Pull request merged**